### PR TITLE
feat(api): list authentication provider configurations

### PR DESCRIPTION
**Pull request by kduret**
## Description

list authentication provider configurations

**Fixes** MON-12527

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

### CODE REVIEW

File: doc/API/centreon-api-v22.04.yaml [hunk 1]
```diff
@@ -146,43 +146,45 @@ paths:
...
+                      description: "Name of provider"
+                      example: "local"
+                    authentication_uri:
+                      type: "string"
+                      description: "Provider Uri used to proceed authentication"
+                      example: "/centreon/authentication/providers/configurations/local"
+                    centreon_base_uri:
+                      type: "string"
+                      description: "Centreon Base Uri used by the provider to redirect"
+                      example: "/centreon"
```
Comment (jeremyjaouen) [inline doc/API/centreon-api-v22.04.yaml:179]: still returned ?

File: src/Core/Application/Security/UseCase/FindProviderConfigurations/FindProviderConfigurations.php [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        \Traversable $providerResponses,
+    ) {
+        if (iterator_count($providerRepositories) === 0) {
+            throw new NotFoundException(_('No provider repositories could be found'));
+        }
+        $this->providerRepositories = iterator_to_array($providerRepositories);
+
+        if (iterator_count($providerResponses) === 0) {
+            throw new NotFoundException(_('No provider responses could be found'));
+        }
```
Comment (jeremyjaouen) [inline src/Core/Application/Security/UseCase/FindProviderConfigurations/FindProviderConfigurations.php:57]: manage translation

File: src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/LocalProviderResponse.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+     * @var bool
+     */
+    public bool $isForced;
+
+    /**
+     * @inheritDoc
+     */
+    public function getType(): string
+    {
+        return 'local';
```
Comment (jeremyjaouen) [inline src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/LocalProviderResponse.php:57]:
```suggestion
        return LocalProvider::NAME;
```

File: src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+     * @var boolean
+     */
+    public bool $verifyPeer;
+
+    /**
+     * @inheritDoc
+     */
+    public function getType(): string
+    {
+        return 'openid';
```
Comment (jeremyjaouen) [inline src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php:117]:
```suggestion
        return OpenIdConfiguration::NAME;
```

File: src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * @var string[]
+     */
+    public array $trustedClientAddresses;
+
+    /**
+     * @var string[]
+     */
+    public array $blacklistClientAddresses;
+
```
Comment (jeremyjaouen) [inline src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php:51]:
```suggestion
```
not needed in response

File: src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * @var string[]
+     */
+    public array $connectionScopes;
+
+    /**
+     * @var string|null
+     */
+    public ?string $loginClaim;
```
Comment (jeremyjaouen) [inline src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php:90]:
```suggestion
```

not needed in response

File: src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    /**
+     * @var string|null
+     */
+    public ?string $authenticationType;
+
+    /**
+     * @var boolean
+     */
+    public bool $verifyPeer;
+
```
Comment (jeremyjaouen) [inline src/Core/Application/Security/UseCase/FindProviderConfigurations/ProviderResponse/OpenIdProviderResponse.php:111]:
```suggestion
```
not needed in response

**Review by jeremyjaouen — CHANGES_REQUESTED** [review]

File: config/packages/Centreon.yaml [hunk 8]
```diff
@@ -221,10 +221,24 @@ services:
                   arguments: ['@sqlRequestTranslator']
         Core\Infrastructure\RealTime\Api\Hypermedia\HypermediaProviderInterface:
             tags: ['realtime.api.hypermedia']
+        Core\Application\Security\ProviderConfiguration\Repository\ReadProviderConfigurationsRepositoryInterface:
+            tags: ['authentication.provider.repositories']
+        Core\Application\Security\UseCase\FindProviderConfigurations\ProviderResponse\ProviderResponseInterface:
+            tags: ['authentication.provider.responses']
+        Core\Infrastructure\Security\Api\FindProviderConfigurations\ProviderPresenter\ProviderPresenterInterface:
+            tags: ['authentication.presenters']
```
Comment (kduret) [inline config/packages/Centreon.yaml:229]:
```suggestion
            tags: ['authentication.provider.presenters']
```

File: config/packages/Centreon.yaml [hunk 9]
```diff
@@ -221,10 +221,24 @@ services:
...
     Centreon\Domain\Monitoring\Interfaces\ResourceServiceInterface:
         class: Centreon\Domain\Monitoring\ResourceService
 
+    Core\Application\Security\UseCase\FindProviderConfigurations\FindProviderConfigurations:
+            arguments:
+              - !tagged_iterator 'authentication.provider.repositories'
+              - !tagged_iterator 'authentication.provider.responses'
+
+    Core\Infrastructure\Security\Api\FindProviderConfigurations\FindProviderConfigurationsPresenter:
+        arguments: [!tagged_iterator 'authentication.presenters']
```
Comment (kduret) [inline config/packages/Centreon.yaml:240]:
```suggestion
        arguments: [!tagged_iterator 'authentication.provider.presenters']
```

File: src/Core/Application/Security/UseCase/FindProviderConfigurations/FindProviderConfigurations.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+        $responses = [];
+        foreach ($configurations as $configuration) {
+            foreach ($this->providerResponses as $providerResponse) {
+                if ($configuration->getType() === $providerResponse->getType()) {
+                    $responses[] = $providerResponse->create($configuration);
+                }
+            }
+        }
```
Comment (kduret) [inline src/Core/Application/Security/UseCase/FindProviderConfigurations/FindProviderConfigurations.php:87]: split in privzte method and describe in phpdoc

Comment (kduret) [reply src/Core/Application/Security/UseCase/FindProviderConfigurations/FindProviderConfigurations.php:87]: or just comment block

File: src/Core/Domain/Security/ProviderConfiguration/OpenId/Model/OpenIdConfiguration.php [hunk 11]
```diff
@@ -28,6 +28,13 @@
 class OpenIdConfiguration
 {
     public const DEFAULT_LOGIN_GLAIM = 'preferred_username';
+    public const TYPE = 'openid';
+    public const NAME = 'openid';
+
+    /**
+     * @var int|null
+     */
+    private ?int $id;
```
Comment (kduret) [inline src/Core/Domain/Security/ProviderConfiguration/OpenId/Model/OpenIdConfiguration.php:37]: Suggestion but not mandatory here : Create NewOpenIdConfiguration model which does not contain `id` property

File: src/Core/Infrastructure/Security/Api/FindProviderConfigurations/ProviderPresenter/LocalProviderPresenter.php [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace Core\Infrastructure\Security\Api\FindProviderConfigurations\ProviderPresenter;
+
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Core\Application\Security\UseCase\FindProviderConfigurations\ProviderResponse\LocalProviderResponse;
+use Core\Infrastructure\Security\Api\FindProviderConfigurations\ProviderPresenter\ProviderPresenterInterface;
+
+class LocalProviderPresenter implements ProviderPresenterInterface
+{
+    public function __construct(private UrlGeneratorInterface $router)
```
Comment (kduret) [inline src/Core/Infrastructure/Security/Api/FindProviderConfigurations/ProviderPresenter/LocalProviderPresenter.php:31]: add phpdoc

File: src/Core/Infrastructure/Security/Api/FindProviderConfigurations/ProviderPresenter/OpenIdProviderPresenter.php [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+namespace Core\Infrastructure\Security\Api\FindProviderConfigurations\ProviderPresenter;
+
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Core\Domain\Security\ProviderConfiguration\OpenId\Model\OpenIdConfiguration;
+use Core\Application\Security\UseCase\FindProviderConfigurations\ProviderResponse\OpenIdProviderResponse;
+use Core\Infrastructure\Security\Api\FindProviderConfigurations\ProviderPresenter\ProviderPresenterInterface;
+
+class OpenIdProviderPresenter implements ProviderPresenterInterface
+{
+    public function __construct(private UrlGeneratorInterface $router)
```
Comment (kduret) [inline src/Core/Infrastructure/Security/Api/FindProviderConfigurations/ProviderPresenter/OpenIdProviderPresenter.php:32]: add phpdoc

File: src/Core/Infrastructure/Security/ProviderConfiguration/Local/Repository/DbConfigurationFactory.php [hunk 14]
```diff
@@ -29,30 +29,41 @@ class DbConfigurationFactory
...
+            $customConfiguration['password_security_policy']['attempts'],
+            $customConfiguration['password_security_policy']['blocking_duration'],
+            $customConfiguration['password_security_policy']['password_expiration_delay'],
             $excludedUserAliases,
-            $configuration['password_security_policy']['delay_before_new_password'],
+            $customConfiguration['password_security_policy']['delay_before_new_password'],
         );
 
-        return new Configuration($securityPolicy);
+        $localConfiguration =  (new Configuration($securityPolicy))
```
Comment (kduret) [inline src/Core/Infrastructure/Security/ProviderConfiguration/Local/Repository/DbConfigurationFactory.php:60]:
```suggestion
        $localConfiguration = (new Configuration($securityPolicy))
```

File: tests/php/Core/Application/Security/UseCase/LoginSession/LoginSessionTest.php [hunk 15]
```diff
@@ -183,7 +183,7 @@ public function testExecuteProviderConfigurationNotFound(): void
             ->willReturn(null);
 
         $this->expectException(ProviderException::class);
-        $this->expectExceptionMessage('Provider configuration (provider_configuration_1) not found');
+        $this->expectExceptionMessage('Provider configuration (local) not found');
```
Comment (kduret) [inline tests/php/Core/Application/Security/UseCase/LoginSession/LoginSessionTest.php:186]: call static method

File: tests/php/Security/Domain/Authentication/ProviderServiceTest.php [hunk 16]
```diff
@@ -310,7 +267,7 @@ public function testFindProviderConfigurationByConfigurationNameFailed(): void
             ->willThrowException(new \Exception());
 
         $this->expectException(ProviderException::class);
-        $this->expectExceptionMessage('Error while searching providers configurations');
+        $this->expectExceptionMessage("Error while searching provider configuration: 'local'");
```
Comment (kduret) [inline tests/php/Security/Domain/Authentication/ProviderServiceTest.php:270]: call static exception method

**Review by kduret — COMMENTED** [review]
check authentication issue when session is expired

File: tests/php/Core/Application/Security/UseCase/FindProviderConfigurations/FindProviderConfigurationsTest.php [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    );
+});
+
+it('returns error when there is an issue during configurations search', function () {
+    $errorMessage = 'error during configurations search';
+
+    $this->readProviderConfigurationRepository
+        ->expects($this->once())
+        ->method('findConfigurations')
+        ->willThrowException(New \Exception($errorMessage));
```
Comment (kduret) [inline tests/php/Core/Application/Security/UseCase/FindProviderConfigurations/FindProviderConfigurationsTest.php:55]:
```suggestion
        ->willThrowException(new \Exception($errorMessage));
```

File: src/Security/SessionAPIAuthenticator.php [hunk 18]
```diff
@@ -148,12 +148,13 @@ function ($userIdentifier) {
      */
     private function getUserAndUpdateSession(string $sessionId): UserInterface
     {
+        $this->authenticationService->deleteExpiredSecurityTokens();
+        $this->sessionRepository->deleteExpiredSession();
```
Comment (jeremyjaouen) [inline src/Security/SessionAPIAuthenticator.php:152]: this should be after isValidToken as isValidToken will check token validity and refresh it if possible (if the expired token is deleted we cannot refresh it)

**Review by callapa — APPROVED** [review]

**Review by a-launois — APPROVED** [review]

**Pull request merged**